### PR TITLE
Global carbon budget in ELM

### PR DESCRIPTION
**Pull request by bishtgautam**
Adds global carbon budget in the ELM.

[BFB]

### CODE REVIEW

**Comment by bishtgautam** [comment]
Things to checks before this PR is ready to be merged:

- [ ] Global carbon budget is correct for a historical run
   - [ ] Instantaneous budget
   - [ ] Daily budget
   - [ ] Monthly budget
   - [ ] Yearly budget
   - [ ] Budgets are correct when the model is restarted

- [ ] Global carbon budget is correct for a transient run
   - [ ] Instantaneous budget
   - [ ] Daily budget
   - [ ] Monthly budget
   - [ ] Yearly budget
   - [ ] Budgets are correct when the model is restarted

- [ ] Implement an error tolerance for the carbon budget

**Comment by susburrows** [comment]
Hi Gautam, thanks for working on this so quickly, and for keeping us in the loop!

**Comment by bishtgautam** [comment]
@susburrows @kvcalvin, This PR is ready for review and integration. Below is the summary of testing that I have done for this PR:

- Simulations have been successfully performed for I1850CNPRDCTCBC I20TRCNPRDCTCBC compsets at f19_g16 resolution
- No modifications were made in `user_nl_clm`.
- ELM BGC includes an option called `ero_ccycle` (probably for erosion cycle). This option is not a default option, so it is not tested.

Please add a reviewer who you think should review/test this PR.

**Comment by susburrows** [comment]
Hi Gautam, thanks for working on this so quickly.  A couple of quick questions:
- What was your basis for selecting the tolerance level of 0.01 kg/m^2 for the global carbon budget?
- How are the budgets being output?  Are they going to the log files?
- What behavior did you implement for when the tolerance check fails?

@thorntonpe: would you perhaps be available and willing to review this PR?

Thank you! Susannah

**Comment by rljacob** [comment]
Notes from IG:  need to fix units on output.

**Comment by bishtgautam** [comment]
@susburrows, Here are answers to your questions:

1. >What was your basis for selecting the tolerance level of 0.01 kg/m^2 for the global carbon budget?

My initial comment regarding global error tolerance had the wrong units. The global error tolerance is `0.01 x 10^6 kg/m^2`. At each patch-level, ELM's carbon balance tolerance is `10^{-8} gC/m^2`. The patch-level carbon error in `gC/m^2` can be converted to `MegaC/m^2`

```
10^{-8} gC/m^2 x 10^-3 kgC/gC x 10^6 MC/kgC= 10^{-6} MC/m^2
```
I added a factor of 10^4 to account for the number of grid cells in the global grid. Thus the global error tolerance is 
```
10^{-6} MC/m^2 * 10^4 = 0.01 MC/m^2
```

2. >How are the budgets being output? Are they going to the log files?

Yes, the budgets are outputted in `*elm.*log` files.

3. >What behavior did you implement for when the tolerance check fails?

The model reports an error and stops.

I do have one outstanding issue regarding budgets that will be reported by ELM and Coupler balance check. ELM budgets are reported in the units of `MC/m^2`, but the data from ELM is sent to the coupler is in the units of `kgCO2/m^2`.

**Comment by susburrows** [comment]
Thanks for the responses, @bishtgautam.  I'll check in with Peter to see if he will be able to have a look at this.

**Comment by rljacob** [comment]
@thorntonpe please review this.

**Comment by bishtgautam** [comment]
I will rebase this PR and ask @peterdschwartz to review+integerate this PR.

**Comment by rljacob** [comment]
waiting for rebase.

**Comment by rljacob** [comment]
@bishtgautam how goes the rebase?

**Comment by bishtgautam** [comment]
@rljacob I have rebased the branch and am testing it to make sure the rebase was done correctly.

**Comment by bishtgautam** [comment]
@rljacob The test run ran overnight on Cori and everything looks good.

@peterdschwartz This PR is ready for your review.

File: components/elm/src/main/elm_driver.F90
Comment (peterdschwartz) [inline components/elm/src/main/elm_driver.F90:402]: Can BeginGridCBalance and BeginGridCBalanceBeforeDynSubgridDriver be combined?  If that's not appealing, perhaps the comment header descriptions should be differentiated.

**Review by peterdschwartz — CHANGES_REQUESTED** [review]

File: components/elm/src/main/elm_driver.F90
Comment (bishtgautam) [reply components/elm/src/main/elm_driver.F90:402]: They can't be combined into one as the different data fields are being used in each subroutine for the conservation check

Comment (peterdschwartz) [reply components/elm/src/main/elm_driver.F90:402]: I'm not sure I follow.  Both routines have the same arguments `bounds, col_cs, grc_cs`  and they are both just calls to `c2g` with no conservation checks.

Comment (bishtgautam) [reply components/elm/src/main/elm_driver.F90:402]: - In [BeginGridCBalance()](https://github.com/E3SM-Project/E3SM/blob/bishtgautam/lnd/cnp-budget/components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90#L808
), there are 8 fields on which `c2g` is performed (e.g. `totcolc`, `totpftc`, etc)

- In [BeginGridCBalanceBeforeDynSubgridDriver](https://github.com/E3SM-Project/E3SM/blob/bishtgautam/lnd/cnp-budget/components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90#L996), there is only one field (`totcolc`) on which `c2g` is performed.

So, there is only one field, `col_cs%totcolc`, that is common between the two subroutines. But, the `g`-level field in which the output of `c2g` is saved are not the same:
  - `BeginGridCBalance`: uses [`grc_cs%beg_totc`](https://github.com/E3SM-Project/E3SM/blob/bishtgautam/lnd/cnp-budget/components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90#L798)
  - `BeginGridCBalanceBeforeDynSubgridDriver`: uses [`grc_cs%begcb`](https://github.com/E3SM-Project/E3SM/blob/bishtgautam/lnd/cnp-budget/components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90#L993)

At some point in the future, we can revisit to merge these two subroutines when we also have corresponding subroutines for N and P (i.e. `BeginGridNBalance`, `BeginGridPBalance`)

Comment (peterdschwartz) [reply components/elm/src/main/elm_driver.F90:402]: All the `c2g` calls output to a different field of `grc_cs` which is why is makes sense to group them together -- especially when they have identical arguments, identical descriptions, and are always called one after the other.  Currently, It appears to be as straightforward as concatenating the two and adding `grc_cs%begcb` to the associate clause of `BeginGridCBalance` (or the other way around).  How will N,P budget developments affect this?

Comment (peterdschwartz) [reply components/elm/src/main/elm_driver.F90:402]: Are the `grc_cs` fields used elsewhere even when the budgets are not being computed?  It may make sense to wrap the `BeginGridCBalance` call  in an if statement so that these computations aren't done unless `do_budgets` is true.

File: components/elm/src/data_types/GridcellDataType.F90
Comment (peterdschwartz) [inline components/elm/src/data_types/GridcellDataType.F90:112]: Searching the rest of the code, `grc_cs%begcb` and `grc_cs%beg_totc` appear to always be the equivalent.  
 `grc_cs%begcb`  is only modified in  [BeginGridCBalanceBeforeDynSubgridDriver](https://github.com/E3SM-Project/E3SM/blob/b39f885b293bb3f44aca85c942c7404d0934e8fd/components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90#L996-L1000) with `c2g`.
Likewise, `grc_cs%beg_totc` is only modified in [BeginGridCBalance](https://github.com/E3SM-Project/E3SM/blob/b39f885b293bb3f44aca85c942c7404d0934e8fd/components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90#L808-L809)  with an equivalent call to `c2g` since `col_cs%totcolc` is not modified in between the two calls.   

So one of the fields looks like it can be deleted.  I think `grc_cs%beg_totc` is a better name and more widely used so I would advocate for deleting `begcb`.

I think the same could be done with `grc_cs%endcb` since `c2g` overwrites the `garr`, and `EndGridCBalanceAfterDynSubgridDriver` would need to be changed to use `grc_cs%end_totc`

**Review by thorntonpe — APPROVED** [review]

**Review by peterdschwartz — APPROVED** [review]

**Comment by jingtao-lbl** [comment]
> Things to checks before this PR is ready to be merged:
> 
> * [x]  Global carbon budget is correct for a historical run
>   
>   * [x]  Instantaneous budget
>   * [x]  Daily budget
>   * [x]  Monthly budget
>   * [x]  Yearly budget
>   * [x]  Budgets are correct when the model is restarted
> * [x]  Global carbon budget is correct for a transient run
>   
>   * [x]  Instantaneous budget
>   * [x]  Daily budget
>   * [x]  Monthly budget
>   * [x]  Yearly budget
>   * [x]  Budgets are correct when the model is restarted
> * [x]  Implement an error tolerance for the carbon budget
>   
>   * A global error tolerance of `0.01*10^6` kg/m^2 has been chosen for the carbon budget.
> 
> cc: @susburrows @kvcalvin

hmmmm, I got an error with this _error_tol_ in a AD spinup mode. Right now I just ignore this and continue my runs...will see how impactful this would be.

> Things to checks before this PR is ready to be merged:
> 
> * [x]  Global carbon budget is correct for a historical run
>   
>   * [x]  Instantaneous budget
>   * [x]  Daily budget
>   * [x]  Monthly budget
>   * [x]  Yearly budget
>   * [x]  Budgets are correct when the model is restarted
> * [x]  Global carbon budget is correct for a transient run
>   
>   * [x]  Instantaneous budget
>   * [x]  Daily budget
>   * [x]  Monthly budget
>   * [x]  Yearly budget
>   * [x]  Budgets are correct when the model is restarted
> * [x]  Implement an error tolerance for the carbon budget
>   
>   * A global error tolerance of `0.01*10^6` kg/m^2 has been chosen for the carbon budget.
> 
> cc: @susburrows @kvcalvin

@bishtgautam hmmmm, I got an error with this _error_tol_ in a AD spinup mode. Right now I just ignore this and continue my runs...will see how impactful this would be.

**Comment by jingtao-lbl** [comment]
@bishtgautam hmmmm, I got an error with this error_tol in a AD spinup mode. Right now I just ignore this and continue my runs...will see how impactful this would be.

**Comment by bishtgautam** [comment]
@jingtao-lbl The carbon budget was only tested for CNP+RD+CTC. Are you using the same configuration?

**Comment by jingtao-lbl** [comment]
> @jingtao-lbl The carbon budget was only tested for CNP+RD+CTC. Are you using the same configuration?

No, I am using CNP+ECA+CENTURY.

**Comment by jingtao-lbl** [comment]
No, I am using CNP+ECA+CENTURY.

**Comment by jingtao-lbl** [comment]
> @jingtao-lbl The carbon budget was only tested for CNP+RD+CTC. Are you using the same configuration?

No, I am using CNP+ECA+CENTURY.

**Comment by bishtgautam** [comment]
It is very likely that certain states or fluxes aren't accounted for in the C-budget for the CNP+ECA+CENTURY. Btw, is the model stopping when it encounters a C mass imbalance or it simply reports an error?

**Comment by jingtao-lbl** [comment]
I see...The model stopped when it encountered a C mass imbalance (CNPBudgetMod.F90, Line 916, abs(time_integrated_flux - state_net_change) > error_tol) and it also reported detailed imbalance error.

**Comment by evasinha** [comment]
I was running a crop simulation and the run failed with the following error message:

```
 time integrated flux =    34233035.2768977
 net change in state  =   -701238.221063285
 error                =    34934273.4979610
 ENDRUN:
 ERROR in CNPBudgetMod.F90 at line 916
```

**Pull request merged**
